src/index.tsx

```diff
@@ -129,51 +136,85 @@ class ApiConnection {
   async _start() {
     while (true) {
       await sleepMsec(17000);
-      this.sendMessage({type: 'keepAlive'});
+      if (!this._isDisconnected) this.sendMessage({type: 'keepAlive'});
     }
   }
 }
-const apiConnection = new ApiConnection();
-const baseSha1Url = `http://${window.location.hostname}:15309/sha1`;
-const hither = initializeHitherInterface((msg) => apiConnection.sendMessage(msg), baseSha1Url)
 
-apiConnection.onConnect(() => {
-  console.info('Connected to API server');
-})
-apiConnection.onMessage(msg => {
-  const type0 = msg.type;
-  if (type0 === 'reportServerInfo') {
-    theStore.dispatch({
-      type: SET_SERVER_INFO,
-      serverInfo: msg.serverInfo
-    });
-  }
-  else if (type0 === 'action') {
-    let action = msg.action;
-    if ('persistKey' in action) {
-      // just to be safe
-      delete action['persistKey'];
+const createApiConnection = () => {
+  const x = new ApiConnection()
+
+  x.onMessage(msg => {
+    const type0 = msg.type;
+    if (type0 === 'reportServerInfo') {
+      theStore.dispatch({
+        type: SET_SERVER_INFO,
+        serverInfo: msg.serverInfo
+      });
     }
-    theStore.dispatch(action);
-  }
-  else if (type0 === 'reportInitialLoadComplete') {
-    theStore.dispatch({
-      type: REPORT_INITIAL_LOAD_COMPLETE
-    });
-  }
-  else if (type0 === 'hitherJobFinished') {
-    hither.handleHitherJobFinished(msg);
-  }
-  else if (type0 === 'hitherJobError') {
-    hither.handleHitherJobError(msg);
-  }
-  else if (type0 === 'hitherJobCreated') {
-    hither.handleHitherJobCreated(msg);
+    else if (type0 === 'action') {
+      let action = msg.action;
+      if ('persistKey' in action) {
+        // just to be safe
+        delete action['persistKey'];
+      }
+      theStore.dispatch(action);
+    }
+    else if (type0 === 'reportInitialLoadComplete') {
+      theStore.dispatch({
+        type: REPORT_INITIAL_LOAD_COMPLETE
+      });
+    }
+    else if (type0 === 'hitherJobFinished') {
+      hither.handleHitherJobFinished(msg);
+    }
+    else if (type0 === 'hitherJobError') {
+      hither.handleHitherJobError(msg);
+    }
+    else if (type0 === 'hitherJobCreated') {
+      hither.handleHitherJobCreated(msg);
+    }
+    else {
+      console.warn(`Unregognized message type from server: ${type0}`)
+    }
+  });
+
+  return x
+}
+
+let apiConnection: ApiConnection
+let queuedHitherJobMessages: HitherJobMessage[] = []
+const handleReconnect = () => {
+  apiConnection = createApiConnection()
```

Comment:
Here's the handler for when user clicks to reconnect. It is also called on initial page load.